### ISSUE TITLE
How to realize pharmacological property prediction through this model

### ISSUE THREAD

**Issue by Li-Xuewei12** [LongLee220/KA-GNN#4]
The AUC performance of this model is very good, but how can we realize the Tox21 prediction for new molecules using the trained model.pth file? I urgently want to know how to do this and look forward to your reply as soon as possible.

**Comment by LongLee220** [comment]
Yes. If you want to use the trained KA-GNN on Tox21 to predict new molecules, you can:
# 1. Directly use the pretrained KA-GNN
Make sure you featurize new molecules exactly the same way as in training (atom/bond features, cutoffs, normalization, and the exact task order of the 12 Tox21 labels).
```
# python
import torch
from torch.utils.data import DataLoader

# 1) Build the same architecture
kan_gat = KAN_GAT(
    in_node_dim=encode_dim[0],
    in_edge_dim=encode_dim[1],
    hidden_dim=64,
    out_1=32,
    out_2=12,   # 12 tasks for Tox21
    gride_size=grid,
    head=head,
    layer_num=num_layers,
    pooling=pooling
)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
state_dict = torch.load("model.pth", map_location=device)
kan_gat.load_state_dict(state_dict)
kan_gat = kan_gat.to(device).eval()

# 2) Featurize new molecules (same pipeline as training)
new_dataset = NewMoleculeDataset(new_graphs)
new_loader  = DataLoader(new_dataset, batch_size=128, shuffle=False, collate_fn=new_collate)

# 3) Inference
all_probs = []
with torch.no_grad():
    for batch in new_loader:
        g = batch.to(device)
        node_x = g.ndata['feat']
        edge_x = g.edata['feat']
        logits = kan_gat(g, node_x, edge_x)   # [B, 12]

        # If trained with BCEWithLogitsLoss -> apply sigmoid here
        probs = torch.sigmoid(logits)
        all_probs.append(probs.cpu())

all_probs = torch.cat(all_probs, dim=0)  # [N, 12]
pred = (all_probs >= 0.5).int()          # optional: binarize
```

# 2. Add a new head on top of KA-GNN
You can wrap the pretrained backbone and add a new head. Freeze the backbone first, train only the new head, then (optionally) unfreeze for fine-tuning.
```
# python
import torch
import torch.nn as nn

class KAN_GAT_New(nn.Module):
    def __init__(self, kagat_model, new_model):
        super().__init__()
        self.base = kagat_model   # pretrained backbone
        self.new  = new_model     # your new head

    def forward(self, g, node_x, edge_x):
        x = self.base(g, node_x, edge_x)
        return self.new(x)

# Load pretrained backbone
kan_gat = KAN_GAT(...)
kan_gat.load_state_dict(torch.load("model.pth", map_location=device))

# Wrap & freeze
model = KAN_GAT_New(kan_gat, new_model).to(device)
for p in model.base.parameters():
    p.requires_grad = False

# Optimizer only updates the new head
optimizer = torch.optim.Adam(
    filter(lambda t: t.requires_grad, model.parameters()),
    lr=1e-3
)
```

## I hope this is helpful to you.

**Comment by Li-Xuewei12** [comment]
Thank you for your kind help. I used the trained model to predict, but noticed that the predicted distribution frequencies all cluster around 0.5. This makes me concerned about the reliability of the predictions. Do you know the reason for this, or could you offer me some suggestions? Look forward to your reply.

<img width="1286" height="430" alt="Image" src="https://github.com/user-attachments/assets/56a2981e-593e-4a45-945a-25f5c86533f8" />

```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Prediction script - Make predictions for new molecules
"""

import torch
import pandas as pd
import numpy as np
from torch.utils.data import Dataset, DataLoader
import dgl
from model.ka_gnn import KA_GNN, KA_GNN_two
from model.mlp_sage import MLPGNN, MLPGNN_two
from model.kan_sage import KANGNN, KANGNN_two
from utils.graph_path import path_complex_mol
import argparse
import yaml


def set_seed(seed):
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False
    np.random.seed(seed)


class PredictionDataset(Dataset):
    def __init__(self, smiles_list, encoder_atom, encoder_bond):
        self.smiles_list = smiles_list
        self.encoder_atom = encoder_atom
        self.encoder_bond = encoder_bond
        self.graphs = []
        self.valid_indices = []

        # Convert SMILES to graph structures
        for i, smiles in enumerate(smiles_list):
            try:
                graph = path_complex_mol(smiles, encoder_atom, encoder_bond)
                if graph and not self.has_node_with_zero_in_degree(graph):
                    self.graphs.append(graph)
                    self.valid_indices.append(i)
                else:
                    print(f"Skipping invalid molecule: {smiles}")
                    self.graphs.append(None)
            except Exception as e:
                print(f"Error processing molecule {smiles}: {e}")
                self.graphs.append(None)

    def has_node_with_zero_in_degree(self, graph):
        return (graph.in_degrees() == 0).any()

    def __len__(self):
        return len(self.smiles_list)

    def __getitem__(self, idx):
        return self.graphs[idx] if self.graphs[idx] is not None else None


def collate_prediction_fn(batch):
    # Filter out None values (invalid graphs)
    valid_graphs = [g for g in batch if g is not None]
    if len(valid_graphs) == 0:
        return None
    batched_graph = dgl.batch(valid_graphs)
    return batched_graph


def message_func(edges):
    return {'feat': edges.data['feat']}


def reduce_func(nodes):
    num_edges = nodes.mailbox['feat'].size(1)
    agg_feats = torch.sum(nodes.mailbox['feat'], dim=1) / num_edges
    return {'agg_feats': agg_feats}


def update_node_features(g):
    g.send_and_recv(g.edges(), message_func, reduce_func)
    g.ndata['feat'] = torch.cat((g.ndata['feat'], g.ndata['agg_feats']), dim=1)
    return g


def load_model(config_path, model_path, target_dim, device):
    """Load trained model"""
    with open(config_path, "r") as config_file:
        config = yaml.safe_load(config_file)

    model_select = config['model_select']
    grid_feat = config['grid_feat']
    num_layers = config['num_layers']
    pooling = config['pooling']
    encoder_atom = config['encoder_atom']
    encoder_bond = config['encoder_bond']

    encode_dim = [92, 21]  # Consistent with training

    # Select model architecture based on configuration
    if model_select == 'ka_gnn':
        model = KA_GNN(in_feat=encode_dim[0] + encode_dim[1], hidden_feat=64,
                       out_feat=32, out=target_dim, grid_feat=grid_feat,
                       num_layers=num_layers, pooling=pooling, use_bias=True)
    elif model_select == 'ka_gnn_two':
        model = KA_GNN_two(in_feat=encode_dim[0] + encode_dim[1], hidden_feat=64,
                           out_feat=32, out=target_dim, grid_feat=grid_feat,
                           num_layers=num_layers, pooling=pooling, use_bias=True)
    elif model_select == 'mlp_sage':
        model = MLPGNN(in_feat=encode_dim[0] + encode_dim[1], hidden_feat=64,
                       out_feat=32, out=target_dim, grid_feat=grid_feat,
                       num_layers=num_layers, pooling=pooling, use_bias=True)
    elif model_select == 'mlp_sage_two':
        model = MLPGNN_two(in_feat=encode_dim[0] + encode_dim[1], hidden_feat=64,
                           out_feat=32, out=target_dim, grid_feat=grid_feat,
                           num_layers=num_layers, pooling=pooling, use_bias=True)
    elif model_select == 'kan_sage':
        model = KANGNN(in_feat=encode_dim[0] + encode_dim[1], hidden_feat=64,
                       out_feat=32, out=target_dim, grid_feat=grid_feat,
                       num_layers=num_layers, pooling=pooling, use_bias=True)
    elif model_select == 'kan_sage_two':
        model = KANGNN_two(in_feat=encode_dim[0] + encode_dim[1], hidden_feat=64,
                           out_feat=32, out=target_dim, grid_feat=grid_feat,
                           num_layers=num_layers, pooling=pooling, use_bias=True)
    else:
        raise ValueError(f"Unknown model type: {model_select}")

    # Load model weights
    model.load_state_dict(torch.load(model_path, map_location=device))
    model.to(device)
    model.eval()

    return model, encoder_atom, encoder_bond


def predict(model, device, data_loader, smiles_list, dataset):
    """Make predictions"""
    model.eval()
    all_predictions = []
    valid_smiles = []

    with torch.no_grad():
        for batch_idx, batched_graph in enumerate(data_loader):
            if batched_graph is None:
                continue

            # Update node features (consistent with training)
            graph_list = update_node_features(batched_graph).to(device)
            node_features = graph_list.ndata['feat'].to(device)

            # Forward propagation
            output = model(graph_list, node_features).cpu()

            # Apply sigmoid to get probabilities
            probabilities = torch.sigmoid(output)

            all_predictions.append(probabilities.numpy())

            # Record corresponding SMILES
            batch_smiles_indices = list(range(batch_idx * data_loader.batch_size,
                                              min((batch_idx + 1) * data_loader.batch_size, len(smiles_list))))
            valid_batch_indices = [i for i in batch_smiles_indices if i in dataset.valid_indices]
            valid_smiles.extend([smiles_list[i] for i in valid_batch_indices])

    if all_predictions:
        all_predictions = np.vstack(all_predictions)
    else:
        all_predictions = np.array([])

    return valid_smiles, all_predictions


def main():
    # Set device
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    print(f'Using device: {device}')

    # Set random seed
    set_seed(42)

    # Parameter settings
    config_path = './config/c_path.yaml'
    model_path = 'model-tox21.pth'  # Trained model file
    data_file = 'train2398.csv'  # Prediction data file
    batch_size = 32
    target_dim = 12  # Modify according to your task, e.g., tox21 is 12

    # Load model
    print("Loading model...")
    model, encoder_atom, encoder_bond = load_model(config_path, model_path, target_dim, device)

    # Load prediction data
    print("Loading prediction data...")
    try:
        df = pd.read_csv(data_file)
        smiles_list = df['smiles'].tolist()
        if 'id' in df.columns:
            id_list = df['id'].tolist()
        else:
            id_list = [f"mol_{i}" for i in range(len(smiles_list))]
        print(f"Successfully loaded {len(smiles_list)} molecules")
    except Exception as e:
        print(f"Failed to load data file: {e}")
        return

    # Create dataset and data loader
    dataset = PredictionDataset(smiles_list, encoder_atom, encoder_bond)
    data_loader = DataLoader(dataset, batch_size=batch_size, shuffle=False,
                             collate_fn=collate_prediction_fn)

    # Make predictions
    print("Starting prediction...")
    valid_smiles, predictions = predict(model, device, data_loader, smiles_list, dataset)

    # Save results
    if len(valid_smiles) > 0:
        # Create result DataFrame
        results_df = pd.DataFrame({
            'smiles': valid_smiles,
            'id': [id_list[smiles_list.index(smiles)] for smiles in valid_smiles]
        })

        # Add prediction result columns
        for i in range(predictions.shape[1]):
            results_df[f'prediction_{i}'] = predictions[:, i]

        # Save to CSV file
        output_file = 'prediction_results.csv'
        results_df.to_csv(output_file, index=False)
        print(f"Prediction completed! Results saved to: {output_file}")
        print(f"Successfully predicted {len(valid_smiles)} molecules")
        print("\nPrediction results for first 5 molecules:")
        print(results_df.head())
    else:
        print("No valid molecules to predict")


if __name__ == '__main__':
    main()
```

**Comment by LongLee220** [comment]
Thank you for your question. Probabilities clustering around 0.5 often suggest model uncertainty. It may help to add more task-specific data and check the loss function and class balance during training.

**Comment by LongLee220** [comment]
Hi, I checked the code. The last layer in KA-GNN (KA-GCN/KA-GAT) already uses a sigmoid, so applying another torch.sigmoid in the prediction script is unnecessary and will squeeze outputs toward ~0.5–0.7. Please remove the extra sigmoid to use the model’s probabilities directly.
```
def predict(model, device, data_loader, smiles_list, dataset):
    """Make predictions"""
    model.eval()
    all_predictions = []
    valid_smiles = []

    with torch.no_grad():
        for batch_idx, batched_graph in enumerate(data_loader):
            if batched_graph is None:
                continue

            graph_list = update_node_features(batched_graph).to(device)
            node_features = graph_list.ndata['feat'].to(device)

            probabilities = model(graph_list, node_features).cpu()

            # this is unnecessary, you may try removing this
            # probabilities = torch.sigmoid(output)

            all_predictions.append(probabilities.numpy())

            batch_smiles_indices = list(range(batch_idx * data_loader.batch_size,
                                              min((batch_idx + 1) * data_loader.batch_size, len(smiles_list))))
            valid_batch_indices = [i for i in batch_smiles_indices if i in dataset.valid_indices]
            valid_smiles.extend([smiles_list[i] for i in valid_batch_indices])

    if all_predictions:
        all_predictions = np.vstack(all_predictions)
    else:
        all_predictions = np.array([])

    return valid_smiles, all_predictions
```

**Comment by Li-Xuewei12** [comment]
Thank you for your patience and thoughtful response to my questions.

**Comment by LongLee220** [comment]
You're welcome
